### PR TITLE
Issue #10:  Logging Enhancements

### PR DESCRIPTION
**Pull request by leftink**
# Summary of Enhancements
- User has full control of logging configuration, i.e. handlers, formatters, levels, etc.
- Added a new level DEEP_DEBUG to allow two different levels of debug messages
- Like the python logging facility, added module level log functions, i.e. critical(...), info(...), etc
- Update SimLog to ensure consistent log naming and easily add keyword arguments into the format record
- Added a ColumnFormatter with several features that is now the base class of the default log formatters.  This formatter will allow users the ability to define their own columns for the logs if they want to.
- Added test_logging to demonstrate the features of the logging facility
# Logging Configuration

The user can now specify a logging configuration file to be loaded at start-up by setting the environment variable **COCOTB_LOG_FMT**.
The user has access to the logging configuration through `cocotb.log.configure(cfg)`, where cfg is one of the formats listed below.
## Configuration Formats
- INI format (see python logging fileConfig format)
  - cfg (str):  Name of file containing configuration
  - cfg (obj): File-Like object, i.e. StringIO containing the configuration
- DICT format (see python logging dictConfig format)
  - cfg (dict): A dictionary containing the configuration
  - cfg (str):  Name of file containing configuration encoded with JSON
  - cfg (obj): File-Like object, i.e. StringIO containing configuration encoded with JSON
  - If PyYAML available:
    - cfg (str):  Name of file containing configuration encoded with YAML
    - cfg (obj): File-Like object, i.e. StringIO containing configuration encoded with YAML

**Note**: dictConfig() was not available until Python 2.7 so the full power of using a dictionary configuration will not work, but I did implement the "incremental" configuration so even older versions of python can benefit from using a DICT configuration for setting logger and handler levels.
# Logging Levels

All existing log levels have been brought into the cocotb.log namespace so users should not need to import the python logging module at all.

Added a new Log Level **DEEP_DEBUG** and registered with the python logging module.  I have always found two levels of debugging to be extremely useful.  An example of where this could be used today would be to overhaul the debugging in the scheduler.  Currently there is another Environment variable to enable debugging in the scheduler, but that could be removed now and then current logging updated to have some basic deubug and then use the **DEEP_DEBUG** level for the information that will print a lot.  With the other changes, this will allow the users to have fine control when to enable/disable this extra logging.

Added a function `register_level_notify_cb(logger, f)` to register a callback function to be notified when the effective level of the logger changes.  An example of this use, is to notify the simulator module when the logger 'cocotb.gpi' changes levels.
## COCOTB Group Log Level Getters/Setters

I added some utility functions into the cocotb namespace to allow users to easily set/get the level for a particular group of loggers.

They take the form of:
`<group>LogSetLevel(lvl)`, e.g. cocotbLogSetLevel(lvl)
`<group>LogGetLevel()`, e.g. cocotbLogGetLevel()

Groups:
- cocotb:      Top Level logger, changes level for all loggers not explicitly set
- gpi:         Controls the level for the GPI interface
- coroutine:   Controls the level for all coroutines not explicitly set
- function:    Controls the level for all "functions" not explicitly set
- driver:      Controls the level for all drivers not explicitly set
- handle:      Controls the level for all handlers not explicitly set
- monitor:     Controls the level for all monitors not explicitly set
- regression:  Controls the level for the RegressionManager
- scheduler:   Controls the level for the Scheduler
- scoreboard:  Controls the level for all scoreboards not explicitly set
# Log Functions

There are module level logging functions that use the 'cocotb' logger.  I have also added a equivalent logging functions for logging deep_debug, i.e. log.deep_debug(...).
# SimLog Enhancements

Log naming no longer requires everyone to specify their logger name as 'cocotb.<name>'.  This just ensures that all logs that are created will fall under the 'cocotb' log hierarchy and therefore get the benefit of the default log level.

Normally for standard logging, if you want to add addtional "fields" to the record that is passed to the formatter, you would have to create a dictionary and pass them through the 'extra' keyword arguments.  The updated SimLog functions will automatically take any keyword arguments that are not standard and automatically pack them into a dictionary and send to the standard "extra" argument.  This is useful to take advantage of several of the features in the formatter described below.
# Column Log Formatter

Logging formatter that formats fields in columns, ensuring the text does not exceed the column width through truncation.  The formatter is highly configurable.  All columns must have a fixed length with the exception of the last column.  The formatter supports multiple line messages with appropriate padding.

The user specifies a list of format strings for both the 'fixed' and the 'optional' columns.  Column formats must be specified in the string format style, e.g. {col:8s}.  The 'fixed' columns will always be formatted so the keys must be present in the record.  The 'optional' columns will try to format each column, but will not error if the key is not present in the record.  For 'optional' columns, if a key is not present in the record, the column is padded.  For 'optional' columns, if any optional column is present, then all optional columns are added to the log, using padded version if necessary, but if none of the optional column keys are present in the record, all of the optional columns will be suppressed.  We have used the optional columns to allow the user to specify an 'id' and a 'label' for each message that is not always present.

The log will have the following format:
`<fixed_1><sep><fixed_2><sep>...<fixed_n><sep>[<optional_1><sep><optional_2><sep>...<sep><optional_n><sep>]<message>`
## Configuration Parameters

```
                   fmt (str): Fromat for the last column
               datefmt (str): Format string for creating {asctime}
            simtimefmt (str): Format string for creating {simtime}
             separator (str): The string separating the columns
                prefix (str): A prefix that applied to the fmt string if formatting works
               divider (int): Length of the divider/header markers
                fixed (list): List of formats for persistent columns
             optional (list): List of formats for optional columns
```
## SimTime Format

The user can specify the format and the resolution for the {simtime} field.  An example of the format string `'{>6.2f}ns'`
## Runtime Controls

The following boolean commands can be added as keyword arguments to the logger function, i.e. logger.info(...), to have some control over formatting by setting them True/False:
- include_optional
- suppress
- header
- divider
### include_optional

This control will force the optional columns to be shown (if True) and hidden (if False) regardless of the default formatting rules.
### suppress

If set to True, will suppress **ALL** columns and only log the 'message' without any columns pre-pended.
### header

If True, will put the 'message' into a header like below:

```
        -----------------------------------------------------------------------------
        -- <message>                                                               --
        -----------------------------------------------------------------------------
```
### divider

If True, will put the message into a divider string and if the message is empty, will just create a divider.

```
        ---------------------------------<message>-----------------------------------
or
        -----------------------------------------------------------------------------
```

### CODE REVIEW

**Comment by ilesser** [comment]
Hi @leftink,

Thank you very much for this! I believe it will be very helpful for debugging certain projects.

What's the status of the PR?
Is there anything I can help with?

**Comment by leftink** [comment]
It is done.  I am just waiting for it to be pulled in.

**Comment by Martoni** [comment]
Still waiting for be pulled in ?

**Comment by imphil** [comment]
@leftink Sorry for the very long delay, the cocotb didn't have sufficient resources to review your patch up to now. This has now changed, and if you're still interested in this one we can start the review process.

I took a bit of time to look through the patches and some things would require more discussion. As a starting point, could you rebase your patches on top of the current master branch to pick up the changes made in the meantime?

Again, sorry for the long delay, I'm sure we can now find common ground for this series and get things merged quickly.

**Comment by leftink** [comment]
@imphil The code has been re-based.  I am happy to discuss whatever.  I made a few updates by the way based on some experience our group has had using the logger.

File: cocotb/log.py [hunk 1]
```diff
@@ -117,106 +699,372 @@ def _printRecord(self, level, filename, lineno, msg, function):
...
+              simtimefmt (str): Format string for creating {simtime}
+               separator (str): The string separating the columns
+           divider_char (char): Character for the divider markers
+             divider_len (int): Length of the divider markers
+            header_char (char): Character for the divider markers
+              header_len (int): Length of the divider markers
+                columns (list): List of formats for persistent columns
+              show_cols (list): List of columns to display (None means all, Empty only shows the message)
+
+        Excetpions:
```
Comment (cmarqu) [inline cocotb/log.py]: Typo: "Exceptions:"

Comment (leftink) [reply cocotb/log.py]: Addressed.

**Comment by leftink** [comment]
@imphil, you asked me to rebase and you needed some more information.  I have done the rebase multiple times, what is the additional information you need for this PR.  I would prefer not continually rebasing this and not have any progress made.

**Comment by imphil** [comment]
Thanks for your update to this PR, and sorry for not getting back sooner. As you might have seen there are a lot of outstanding PRs in this repository currently. We're working hard to reduce the backlog, but it will require a bit more time. The logging changes in this PR require a bit more discussion and I'll get back to them as soon as I have a bit more bandwidth for it.

File: cocotb/__init__.py [hunk 2]
```diff
@@ -209,3 +210,162 @@ def process_plusargs():
...
+    }
+    log.configure(cfg)
+
+def _getLevel(name=None):
+    """Returns the effective level of the a logger.
+
+    Args:
+        name (str): name of the logger
+    """
+    return log.SimLog(name=name).getEffectiveLevel()
```
Comment (eric-wieser) [inline cocotb/__init__.py]: I don't really understand the purpose of these functions - can't we just expose the builtin logger objects directly to the user?

**Comment by eric-wieser** [comment]
This conflicts quite heavily - in particular with gh-984, which removed a lot of the cocotb-specific logging infrastructure in favor of falling back on the builtins.

File: cocotb/log.py [hunk 3]
```diff
@@ -30,79 +30,661 @@
...
+                logger = logging.getLogger(name)
+                propagate = cfg.get('propagate', None)
+                if propagate is not None:
+                    logger.propagate = propagate
+                level = cfg.get('level', None)
+                if level is not None:
+                    logger.setLevel(logging._checkLevel(level))
+            except Exception as e:
+                raise ValueError('Unable to configure logger '
+                                 '%r: %s' % (name, e))
```
Comment (eric-wieser) [inline cocotb/log.py]: All such older versions are EOL - this code can be deleted

File: cocotb/log.py [hunk 4]
```diff
@@ -103,12 +103,19 @@ def initialize(top):
...
     _top_logger = SimLog()
 
+    hdlr = logging.StreamHandler(sys.stdout)
+
+    if allow_ansi():
+        hdlr.setFormatter(SimColourLogFormatter())
+    else:
+        hdlr.setFormatter(SimLogFormatter())
+
+    _top_logger.addHandler(hdlr)
```
Comment (eric-wieser) [inline cocotb/log.py]: This change made it in in #1266 :)

**Comment by imphil** [comment]
@eric-wieser @themperek 

I wonder if we can get a 80/20 solution in place that covers the needs of @leftink while leaving some of the special-case stuff to the testbench author to configure (as long as appropriate extension mechanisms are in place).

There's is no point in reviewing this PR in detail if we don't have a clear agreement on what kind of change we're willing to accept.

**Comment by eric-wieser** [comment]
Going through the high-level summary item-by-item:

---

> User has full control of logging configuration, i.e. handlers, formatters, levels, etc.

This is mostly done as of #1266. It might be nice to add a global flag to disable cocotb from enabling its logging in the first place, or some global function to call to undo the cocotb setup. #1225 may ease this too.

> Added a new level DEEP_DEBUG to allow two different levels of debug messages

I don't think we need this in cocotb, the python `logging` module lets users invent whatever levels they like anyway. Typically creating distinct loggers is better than creating new levels anyway

> Like the python logging facility, added module level log functions, i.e. critical(...), info(...), etc

I also don't think we need this, if the user wants these they can just use `logging.info(...)`, and it will log through the same mechanism anyway.

> Update SimLog to ensure consistent log naming and easily add keyword arguments into the format record

I need to look at this some more, but I'm not sure how much value it really adds

> Added a ColumnFormatter with several features that is now the base class of the default log formatters. This formatter will allow users the ability to define their own columns for the logs if they want to.

This seems like a reasonable refactor, but would be a lot easier to review if it stood alone

> Added test_logging to demonstrate the features of the logging facility

Obviously a good thing!

---

One other thing I noticed looking at the diff - this adds hooks to the python logging subclass in order to update the internal gpi log level. Today, setting the python log level will not actually affect the log level used by the C code.

**Comment by imphil** [comment]
@eric-wieser thanks for the summary. Can you take the lead on creating issues for the things we might want to fix, and then close this PR? It's unlikely we'll take it in the form it currently has.

**Pull request closed without merging**